Add improved tracking to the marketplace

kdevnel opened the pull request:
### Changes proposed in this Pull Request:

Closes https://github.com/Automattic/woocommerce.com/issues/18294

The issue linked provides improvements to the interaction that are tracked in the marketplace page. The initial priority is to ensure that all the tracking that currently exists in the PHP marketplace, is still present in the React marketplace. This PR handles that and adds some improvement to that tracking to better suit the new marketplace structure.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

1. Checkout this branch and build the project
2. Visit any WC Admin page such as `/wp-admin/admin.php?page=wc-admin` ([wp-env link](http://localhost:8888/wp-admin/admin.php?page=wc-admin))
3. Open the browser console, the paste and run the following: `localStorage.setItem( 'debug', 'wc-admin:*' );` to allow tracks events from React to show in the console.
4. Visit the marketplace pages and see the various events are recording correctly to the console. They should make sense based on the tables below. Note: the legacy events are mapped to the logic of the events listed [here](https://github.com/woocommerce/woocommerce/blob/9a0e91477eadc4b4a6027197edb4a5ce1d48548c/plugins/woocommerce/includes/tracks/events/class-wc-extensions-tracking.php#L31-L48).
5. Particularly, check for doubling of event triggering that may skew data.
6. Test around the issue

**Legacy events structure**
| Event | Name | Properties |
|--------|--------|--------|
| Discover tab | `extensions_view` | `section: '_featured'` `version: '2'`|
| Browse (extensions) tab | `extensions_view` |  `section: '_all'` `version: '2'` |
| Browse tab with category selected | `extensions_view` |  `section: category name` `version: '2'` |
| Themes tab | `extensions_view` |  `section: 'themes'` `version: '2'` |
| Themes tab with category selected | no event since this isn't currently supported | N/A |
| Search results | `extensions_view_search` |  `section: 'search'` `search_term: search query entered` `version: "2"` |
| Search results > View All button | no event since this isn't currently supported | N/A |
| Search results > View All button with category | no event since this isn't currently supported | N/A |
| Subscriptions tab | `subscriptions_view` |  `section: 'helper'` `version: '2'` |

**New events structure**
Note, all events have the same name – `marketplace_v2_view` so the name column isn't shown.
| Event | Properties |
|--------|--------|
| Discover tab | `view: 'discover'` |
| Browse (extensions) tab | `view: 'extensions'` |
| Browse tab with category selected | `view: extensions` `category: category name` |
| Themes tab | `view: 'themes'` |
| Themes tab with category selected | `view: 'themes'` `category: category name` |
| Search results | `view: 'search'` `searchTerm: search query entered` |
| Search results > View All button | `view: 'search'` `searchTerm: search query entered` `productType: 'extension' or 'theme'` |
| Search results > View All button with category | `view: 'search'` `searchTerm: search query entered` `productType: 'extension' or 'theme'` `category: category name` |
| Subscriptions tab | `view: 'my-subscriptions'` |

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [x] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

Ensure new marketplace works with the existing tracking events.

</details>

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -27,6 +31,18 @@ export default function Content(): JSX.Element {
+		// they're separate methods to make it easier to remove the legacy one later
+		const marketplaceViewProps = {
+			view: selectedTab,
+			searchTerm: query?.term,
+			productType: query?.section,
```
Tracks doesn't like camel case – it compares against [this](https://github.com/woocommerce/woocommerce/blob/f05e8efc17683b56d4941adb702baa16fe8e8cf5/plugins/woocommerce/includes/tracks/class-wc-tracks-event.php#L26) regex and throws error `A valid prop name must be specified`.

```suggestion
			search_term: query?.term,
			product_type: query?.section,
```

kdevnel replied:
Ah, good point! I'll rework things to fit snake case then.

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/products/products.tsx:
```diff
@@ -57,6 +58,11 @@ export default function Products( props: ProductsProps ): JSX.Element {
 	function showSection( section: ProductType ) {
+		recordMarketplaceView( {
+			view: 'search',
+			searchTerm: query?.term,
+			productType: section,
```
```suggestion
			search_term: query?.term,
			product_type: section,
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx:
```diff
@@ -0,0 +1,77 @@
+
+interface MarketplaceViewProps {
+	view: string;
+	searchTerm?: string;
+	productType?: string;
```
```suggestion
	search_term?: string;
	product_type?: string;
```

andfinally reviewed (changes requested):
We need to change this PR to track trunk – we won't be merging the subscriptions feature branch yet!

andfinally commented on plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx:
```diff
@@ -0,0 +1,77 @@
+function recordMarketplaceView( props: MarketplaceViewProps ) {
+	const eventProps = { ...props };
+
+	recordEvent( 'marketplace_v2_view', eventProps );
```
I would be comfortable with calling this `marketplace_view`, as it won't be too hard to find out when we introduced it.

andfinally reviewed (commented):
I've finally figured out how tracks events are sent - `https://stats.wp.com/w.js` makes a request like `https://pixel.wp.com/t.gif?...` 

When the properties aren't present they are showing up as `'undefined'` in the pixel request – I wonder if we should default them to empty string? For example, on http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions.

<img width="156" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/059e50b1-91ff-4e1a-bc56-a8c4bbe83c26">

To view this, I commented out this bit in [packages/js/tracks/src/index.ts](https://github.com/woocommerce/woocommerce/blob/30151a6cf789c8276033e900973837598b660450/packages/js/tracks/src/index.ts#L52-L55):

```js
	if ( isDevelopmentMode ) {
		// window.wcTracks.validateEvent( eventName, eventProperties );
		// return false;
	}
```

kdevnel commented:
> We need to change this PR to track trunk – we won't be merging the subscriptions feature branch yet!

Yeah, that's true @andfinally. I'll need to rebase this, too, so we don't incidentally bring in parts of the feature branch when we merge! I'll work on that.

mlaetitia reviewed (changes requested):
Same issue as Gareth - this is what I see in my console:

![Xnip2023-10-24_10-27-57](https://github.com/woocommerce/woocommerce/assets/30905719/5d8c7676-d550-4b6e-b6f2-156dac2b8b76)

kdevnel commented:
> Same issue as Gareth - this is what I see in my console:
> 
> ![Xnip2023-10-24_10-27-57](https://user-images.githubusercontent.com/30905719/277613645-5d8c7676-d550-4b6e-b6f2-156dac2b8b76.jpg)

This is what happens when you filter errors in your console and forget to turn them back on 😵‍💫

andfinally commented:
> I was actually looking for a concise way to not set these properties at all when they're undefined, but I couldn't for the life of me figure out a way to do it. 

This in `content.tsx` does the trick for me:

```js
		const marketplaceViewProps = {
			view: selectedTab || '',
			search_term: query?.term || '',
			product_type: query?.section || '',
			category: query?.category || '',
		};
```

<img width="135" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/f29c597c-3de6-478f-ae59-da1525b3396b">

andfinally commented on plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	if ( props.view === 'extensions' ) {
+		oldEventName = 'extensions_view';
+
+		oldEventProps.section = props?.category;
```
```suggestion
		oldEventName = 'extensions_view';
		oldEventProps.section = props?.category;
```

kdevnel committed:
Add tracks events for to ensure legacy funnels work

kdevnel commented:
Okay, I addressed all PR feedback and I've rebased on `trunk` now. I am working on additional tracking but I'll likely do those as a separate PR so we can focus on getting the core tracking in place.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

andfinally reviewed (commented):
Thanks for the tweaks, looking nice!

If it's easy, it would be neat to omit a property if it has an empty value. Sorry if I helped create this situation with my earlier suggestion!

I also noticed that viewing say http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions clocks up two pairs of events. Maybe when all properties are empty we can assume the event is just triggered by the initial mount, and skip it?

<img width="1114" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/9eaeb2cd-2716-4fb4-9ef9-cd5c2bb30f49">

Also... events are showing up in Tracks live view – yay!

kdevnel committed:
Fix some logic errors and tidy up empty properties

Removed default values from `marketplaceViewProps` – we are now handling this in the tracking function.

Changed `selectedTab` dependency of `useEffect` hook in content.tsx to `query?.tab`. The `selectedTab` state was being reset after (or at least not at the same time as) the change of the tab param in the query string. For this reason, we were recording two page views when you changed from Extensions tab and Shipping category to Themes tab. This doesn't seem to break anything! We are still setting the `selectedTab` based on the query string in the `Tabs` component, so they should still be in sync.

Removed the condition skipping marketplace page view when there was no `prop.view`. This meant that when you switched to the Discovery tab from another tab, we weren't recording the page view.

Also setting default category prop `_all` when no category is specified on extensions, themes and search tabs. This allows us to bundle the default views of those tabs with the views got by clicking on the "All" category link.

Addressing linter errors.

kdevnel commented:
@andfinally I've made all your suggested changes. I also happened to find a bug that was causing “View All” to trigger 2 events for the new event type and also trigger an old event with the same name as a normal search. I stopped both of those, so we don't artificially inflate event counts.

Also found a couple of nice ways to handle conditionally assigning properties.

mlaetitia reviewed (commented):
On my end it seems that all the legacy events are getting properly recorded but the new ones aren't.
It seems that only category filters are getting recorded on the new event - is this intended?

andfinally reviewed (commented):
I pushed a few small changes in 1f67eb8d5c20748655a0173b3a3aca1c45163f96 which I hope will fix the problem of multiple page view events without breaking anything else! 🙏 The `useEffect` hook we're using to track page views was running twice when the URL changed and the `selectedTab` state changed in response. So I've changed the hook to depend on the tab in the query string instead.

We still have "View All" bug in search results you mentioned – maybe I've reintroduced it, heh. The debugger suggests this is happening because after the first render, we change the `section` query string param. This is the hook's fifth dependency. But we can worry about that another time.

<img width="911" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/68178dcc-c2cf-45d4-8d01-99ce4c3fc167">

I'll give this another go tomorrow!

andfinally committed:
Removed `recordMarketplaceView` from `showSection` in `Products` component. We already record the view in the `useEffect` hook  in the `Content` component, so we were getting a duplicate view.

Only setting default category `_all` for `search` view if `product_type` (i.e. `query.section`) is also set. The previous arrangement was setting the `_all` category on the initial search results, where category doesn't figure.

Removed `! props.view` from condition in `recordLegacyTabView`. It meant that the initial default marketplace view wasn't being recorded.

Solve issue with some default prop values

We were no longer sending through a property to determine the "view" or "section" (depending on which event version you're referring to) when loading the discover page.

This change ensures that there is always a fallback that resolves this.

Also fixed some linting issues and cleaned up redundant code.

Removed unused import.

andfinally reviewed (approved):
I pushed a few more small changes in [`6802488` (#40951)](https://github.com/woocommerce/woocommerce/pull/40951/commits/680248836cfa152b321d8d48c61f457e31e14318).

* Removed `recordMarketplaceView` from `showSection` in `Products` component. We already record the view in the `useEffect` hook  in `Content`. This fixes the duplication with that view.
* Only setting default category `_all` for `search` view if `product_type` (i.e. `query.section`) is also set. The previous method was setting the `_all` category on the initial search results, where category doesn't appear.
* Removed `! props.view` from condition in `recordLegacyTabView`. It meant we weren't recording the legacy view for the initial default marketplace page.

Everything seems to be working OK! @mlaetitia, hopefully you'll now see both sets of events.

kdevnel commented:
Okay, one final change I believe. @mlaetitia you should be able to use the table in the PR description to confirm what events should fire and when. Just to note - run the build and then leave the marketplace page and return. This can sometimes help. I'm not sure if the tracks event logging caches something that causes issues occasionally.

andfinally reviewed (approved):
Thanks for the latest changes! Everything's working nicely.

Pushed a small change to remove an unused import in [`58fe816` (#40951)](https://github.com/woocommerce/woocommerce/pull/40951/commits/58fe816b8fbc4bbce22924eefdee767d198653e2).

Before we merge, it might be nice to squash the commits to one, or do an interactive rebase to squash the small ones.

Dan-Q reviewed (approved):
Works nicely. Love all the comments.

`plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx` doesn't need to be a `.tsx` file. It could just be a `.ts` and therefore use a simpler Babel module to compile it. Not a big deal though!

Love this.

mlaetitia reviewed (approved):
All working now! Thanks !
